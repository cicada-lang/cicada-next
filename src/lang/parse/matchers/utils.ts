import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
