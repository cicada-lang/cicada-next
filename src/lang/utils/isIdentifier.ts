const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
