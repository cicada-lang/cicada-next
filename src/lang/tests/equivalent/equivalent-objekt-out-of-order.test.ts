import { expect, test } from "bun:test"
import { runCode } from "../utils"

test("equivalent Objekt -- out of order", async () => {
  const output = await runCode(`

function id(T: Type, x: T): T {
  return x
}

compute equivalent class { A: Type, x: A, y: A } {
    { A: String, x: "abc", y: "xyz" }
  = { A: String, y: "xyz", x: "abc" }
  = { A: String, x: id(String, "abc"), y: id(String, "xyz") }
  = { A: String, y: id(String, "xyz"), x: id(String, "abc") }
}

`)

  expect(output).toMatchSnapshot()
})
