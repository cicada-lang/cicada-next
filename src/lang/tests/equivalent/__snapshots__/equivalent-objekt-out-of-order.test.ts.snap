// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`equivalent Objekt -- out of order 1`] = `
"refl(
  implicit class {
    A: Type
    x [rename: x5]: A
    y: A
  },
  implicit {
    A: String,
    x: "abc",
    y: "xyz"
  },
): Equal(
  class {
    A: Type
    x [rename: x5]: A
    y: A
  },
  {
    A: String,
    x: "abc",
    y: "xyz"
  },
  {
    A: String,
    x: "abc",
    y: "xyz"
  },
)"
`;
